README.md
<img src="man/figures/logo.png" width = 120 alt="nortsTest Logo"/>

**nortsTest: An R Package for Assessing Normality of Stationary Process**
======================================================================

**nortsTest** is an R package for assessing normality of stationary
process, it tests if a given data follows a stationary Gaussian process.
The package works as an extension of the **nortest** package that performs 
normality tests in random samples (*independent data*). The four principal 
package’s functions are:

-   epps.test() function that implements the [Epps test](https://projecteuclid.org/euclid.aos/1176350618),

-   lobato.test() function that implements the Lobato and Velasco’s test,

-   vavra.test() function that implements the [Psaradaki and Vavra’s test](http://www.applied-econometrics.com),

-   rp.test() function that implements the random projections test of
    Nieto-Reyes, Cuesta-Albertos and Gamboa’s test.

Additionally, inspired in the function *check.residuals()* of the
[forecast package](https://github.com/robjhyndman/forecast), we
provide the **check\_residuals** methods for checking model’s
assumptions using the estimated residuals. The function checks
stationarity, homoscedasticity and normality, presenting a report of the
used tests and conclusions.

Checking normality assumptions
------------------------------

``` r
library(nortsTest)
```

Classic hypothesis tests for normality such as *Shapiro & Wilk, Anderson
& Darling*, or *Jarque & Bera*, do not perform well on dependent data.
Therefore, these tests should not be used to check whether a given time
series has been drawn from a Gaussian process. As a simple example, we
generate a stationary ARMA(1,1) process simulated using an t student
distribution with 7 degrees of freedom, and perform the Anderson-Darling
test from the *nortest package*.

``` r
x = arima.sim(100,model = list(ar = 0.32,ma = 0.25),rand.gen = rt,df = 7)

nortest::ad.test(x)
#> 
#>  Anderson-Darling normality test
#> 
#> data:  x
#> A = 0.50769, p-value = 0.1954
```

The null hypothesis is that the data has a normal distribution and
therefore, follows a Gaussian Process. At *α* = 0.05 significance level
the alternative hypothesis is rejected and wrongly concludes the data
follows a Gaussian process. Applying the Lobato and Velasco’s test of
our package, the null hypothesis is correctly rejected.

``` r
lobato.test(x)
#> 
#>  Lobatos and Velascos test
#> 
#> data:  x
#> lobato = 16.864, df = 2, p-value = 0.0002177
#> alternative hypothesis: x does not follow a Gaussian Process
```

Example: stationary AR(2) process
---------------------------------

In the next example we generate a stationary AR(2) process, using an
exponential distribution with rate of 5, and perform the *epps* and *rp*
with k = 5 random projections tests. With a significance level at
*a**l**p**h**a* = 0.05, the null hypothesis of non-normality is
rejected.

``` r
set.seed(298)
# Simulating the AR(2) process
x = arima.sim(250,model = list(ar =c(0.2,0.3)),rand.gen = rexp,rate = 5)

# tests
epps.test(x)
#> 
#>  epps test
#> 
#> data:  x
#> epps = 38.158, df = 2, p-value = 5.178e-09
#> alternative hypothesis: x does not follow a Gaussian Process
rp.test(x,k = 5)
#> 
#>  k random projections test
#> 
#> data:  x
#> k = 5, lobato = 188.771, epps = 28.385, p-value = 0.0007823
#> alternative hypothesis: x does not follow a Gaussian Process
```

Checking model’s assumptions: cardox data
-----------------------------------------

As an example, we analyze the monthly mean carbon dioxide (*in ppm*)
from the **astsa package**, measured at Mauna Loa Observatory, Hawaii.
from March, 1958 to November 2018. The carbon dioxide data measured as 
the mole fraction in dry air, on Mauna Loa constitute the longest record 
of direct measurements of CO2 in the atmosphere. They were started by C. 
David Keeling of the Scripps Institution of Oceanography in March of 1958 
at a facility of the National Oceanic and Atmospheric Administration.

``` r
library(astsa)
data("cardox")

autoplot(cardox,xlab = "years",ylab = " CO2 (ppm)",color = "darkred",
size = 1,main = "Carbon Dioxide Levels at Mauna Loa")
```

<img src="man/figures/unnamed-chunk-5-1.png" width="60%" style="display: block; margin: auto;" />

The time series clearly has trend and seasonal components, for analyzing
the *cardox* data we proposed a Gaussian linear state space model. We
use the model’s implementation from the [forecast package](https://github.com/robjhyndman/forecast) as follows:

``` r
library(forecast)
#> 
#> Attaching package: 'forecast'
#> The following object is masked from 'package:astsa':
#> 
#>     gas

model = ets(cardox)
summary(model)
#> ETS(M,A,A) 
#> 
#> Call:
#>  ets(y = cardox) 
#> 
#>   Smoothing parameters:
#>     alpha = 0.5591 
#>     beta  = 0.0072 
#>     gamma = 0.1061 
#> 
#>   Initial states:
#>     l = 314.6899 
#>     b = 0.0696 
#>     s = 0.6611 0.0168 -0.8536 -1.9095 -3.0088 -2.7503
#>            -1.2155 0.6944 2.1365 2.7225 2.3051 1.2012
#> 
#>   sigma:  9e-04
#> 
#>      AIC     AICc      BIC 
#> 3136.280 3137.140 3214.338 
#> 
#> Training set error measures:
#>                     ME     RMSE       MAE         MPE       MAPE      MASE
#> Training set 0.0232403 0.312003 0.2430829 0.006308831 0.06883992 0.1559102
#>                    ACF1
#> Training set 0.07275949
```

The best fitted model is a *multiplicative level, additive trend and
seasonality* state space model. If the model’s assumptions are
satisfied, then the model’s errors behave like a Gaussian stationary
process. These assumptions can be checked using our *check\_residuals*
functions.

In this case, we use an Augmented Dickey-Fuller test for stationary
assumption, and a random projections test for normality.

``` r
check_residuals(model,unit_root = "adf",normality = "rp",plot = TRUE)
#> 
#>  *************************************************** 
#> 
#>  Unit root test for stationarity: 
#> 
#>  Augmented Dickey-Fuller Test
#> 
#> data:  y
#> Dickey-Fuller = -9.7249, Lag order = 8, p-value = 0.01
#> alternative hypothesis: stationary
#> 
#> 
#>  Conclusion: y is stationary
#>  *************************************************** 
#> 
#>  Goodness of fit test for Gaussian Distribution: 
#> 
#>  k random projections test
#> 
#> data:  y
#> k = 2, lobato = 3.8260, epps = 1.3156, p-value = 0.3328
#> alternative hypothesis: y does not follow a Gaussian Process
#> 
#> 
#>  Conclusion: y follows a Gaussian Process
#>  
#>  ***************************************************
```

<img src="man/figures/unnamed-chunk-7-1.png" width="60%" style="display: block; margin: auto;" />

Now that all the model’s assumptions are checked, the model is accepted
and can be used to forecast.

``` r
autoplot(forecast(model,h = 12),include = 100,xlab = "years",ylab = " CO2 (ppm)",
         main = "Forecast: Carbon Dioxide Levels at Mauna Loa")
```

<img src="man/figures/unnamed-chunk-8-1.png" width="60%" style="display: block; margin: auto;" />

How to install nortsTest?
-------------------------

The current development version can be downloaded from GitHub via

``` r
if (!requireNamespace("remotes")) install.packages("remotes")

remotes::install_github("asael697/nortsTest",dependencies = TRUE)
```

Additional test functions
-------------------------

The **nortsTest** package offers additional functions for descriptive
analysis in univariate time series.

-   *uroot.test*: performs unit root test for checking stationary in
    linear time series. The Ljung-Box, Augmented Dickey-Fuller,
    Phillips-Perron and Kpps tests can be selected with the *unit\_root*
    option parameter.

-   *seasonal.test*: performs seasonal unit root test for stationary in
    seasonal time series. The hegy, ch and ocsb tests are available with
    the seasonal option parameter.

-   *arch.test*: for checking the ARCH effect in time series. The
    Ljung-Box and Lagrange Multiplier tests can be selected from the
    *arch* option parameter.

-   *normal.test*: for normal distribution check in time series and
    random samples. The tests presented above can be chosen for
    stationary time series. For random samples (*independent data*), the
    Anderson & Darling, Shapiro & Wilks, and Jarque-Bera tests are
    available with the normality option parameter.

For visual diagnostic, we offer ggplot2 methods for numeric and
time-series data. Most of the functions were adapted from Rob Hyndman’s
[forecast package](https://github.com/robjhyndman/forecast).

-   *autoplot*: For plotting time series objects (*ts class*).

-   *gghist*: histograms for numeric and univariate time series.

-   *ggnorm*: quantile-quantile plot for numeric and univariate time
    series.

-   *ggacf & ggpacf*: partial and auto correlation functions plots for
    numeric and univariate time series.

-   *check\_plot*: summary diagnostic plot for univariate starionary
    time series.

Accepted models for residual check
----------------------------------

Currently our check\_residuals() and check\_plot() methods are valid for
the current models and classes:

-   **ts**: for uni variate time series

-   **numeric**: for numeric vectors

-   **arima0**: from the stats package

-   **Arima**: from the forecast package

-   **fGARCH**: from the fGarch package

-   **lm**: from the stats package

-   **glm**: from the stats package

-   **Holt and Winters**: from the stats and forecast package

-   **ets**: from the forecast package

-   **forecast methods**: from the forecast package.

For overloading more functions, methods or packages, please make a pull
request or send a mail to: asael\_am@hotmail.com

References
----------

-   Epps, T.W. (1987). Testing that a stationary time series is
    Gaussian. *The Annals of Statistic*. 15(4), 1683-1698.

-   Nieto-Reyes, A., Cuesta-Albertos, J. & Gamboa, F. (2014). A
    random-projection based test of Gaussianity for stationary
    processes. *Computational Statistics & Data Analysis, Elsevier*.
    75(C), 124-141.

-   Lobato, I., & Velasco, C. (2004). A simple test of normality for
    time series. *Journal Econometric Theory*. 20(4), 671-689.

-   Psaradakis, Z. & Vavra, M. (2017). A distance test of normality for
    a wide class of stationary process. *Journal of Econometrics and
    Statistics*. 2, 50-60.

-   Hyndman, R. & Khandakar, Y. (2008). Automatic time series
    forecasting: the forecast package for R. *Journal of Statistical
    Software*. 26(3), 1-22.


NEWS.md
**Packages NEWS and Updates**
============

**nortsTest 1.0.1 Date: 17/06/2021**
----------------------------------

### Features:

- Add GPL-2 license.

### Changes:

- False discovery rate set as default for the random projections test

- Change "NA %in$ y" to "anyNA(y)"

- use match.args() function for more flexibility

### Fixes:

- Bug fix for the False discovery rate.


**nortsTest 1.0.0 Date: 08/07/2020**
----------------------------------

### Fixes:

- Bug fix for CRAN submission.


**nortsTest 1.0.0 Date: 12/06/2020**
----------------------------------

### Fixes:

- Bug fix for CRAN submission.


**nortsTest 1.0.0 Date: 09/05/2020**
----------------------------------

### Features:

- vavra.test() function for the Psaradaki and Vavra test.

- sieve.bootstrap() function for bootstrap sub-sample in stationary time series.

- vavra.sample() function for the Anderson Darling sample statistics for the Psaradaki and Vavra test.

- rp.test() function for the random projections test.

- rp.sample() function for the random projections statistics samples.

- random.projection() function to generate a random projection of a stationary process.

- gghist(), ggnorm(), ggacf() and ggpacf() function for visualization.

### Improvements:

- epps.statistic() bug fixed.

- package documentation, errors, warnings and notes corrected.

### Changes:

- plot.compare() function deleted

- epps.statistic() using the **poweR** package deleted.

### Fixes:

-   The amoebam algorithm for epps.statistc() is corrected.


**normality 0.0.1.000 Date: 12/03/2020**
----------------------------------

### Features:

-  normal.test() function  with all the normality test available

-  uroot.test() function with all the unit root test available

-  seasonal.test() function with all the seasonal unit root test available

-  arch.test() function with the arch effect test for stationarity

- check_plot() methods implemented for residual diagnostics visualization for lm, glm, ets, forecast, Arima, arima0, fgarch, HoltWinters, ts, and numeric classes

- check_residuals() methods implemented for residual diagnostics for lm, glm, ets, forecast, Arima, arima0, fgarch, HoltWinters, ts, and numeric classes

- the autoplot() methods are overloaded for plotting time series (*ts*) and multivariate (*mts*) classes

### Improvements:

-   lobato.test() is implemented as a htest method

-   epps.test() is implemented as a htest method

### Changes:

-   LV.statistic() function change to lobato.statistic()

### Fixes:

-   The amoebam algorithm for Epps.statistc() is changed for the one implemented by the **PoweR** package


**normality 0.0.1.000 Date: 23/10/2020**
----------------------------------

### Features:

-  Updated the epps.statistic function

-  Updated the LV.statistic function

-  Perform documentation and package description


**normality 0.0.1.000 Date: 22/10/2020**
----------------------------------

### Features:

-  Package Repository created

-  Discussion of the previous work and code homogeneity check README.md

-  Incorporation of references and latex thesis algorithm

-  Incorporation of the article structure


R/Lm_test.R
#' The Lagrange Multiplier test for arch effect.
#'
#' Performs the Lagrange Multipliers test for homoscedasticity in a stationary process.
#' The null hypothesis (H0), is that the process is homoscedastic.
#'
#' @usage  Lm.test(y,lag.max = 2,alpha = 0.05)
#'
#' @param y a numeric vector or an object of the \code{ts} class containing a stationary time series.
#' @param lag.max an integer with the number of used lags.
#' @param alpha Level of the test, possible values range from 0.01 to 0.1. By default
#' \code{alpha = 0.05} is used.
#'
#' @return a h.test class with the main results of the Lagrage multiplier hypothesis test.
#' The h.test class have the following values:
#' \itemize{
#'  \item{"Lm"}{The lagrange multiplier statistic}
#'  \item{"df"}{The test degrees freedoms}
#'  \item{"p.value"}{The p value}
#'  \item{"alternative"}{The alternative hypothesis}
#'  \item{"method"}{The used method}
#'  \item{"data.name"}{The data name.}
#' }
#'
#' @details
#' The Lagrange Multiplier test proposed by \emph{Engle (1982)}
#' fits a linear regression model for the squared residuals and
#' examines whether the fitted model is significant. So the null
#' hypothesis is that the squared residuals are a sequence of
#' white noise, namely, the residuals are homoscedastic.
#'
#' @export
#'
#' @author A. Trapletti and Asael Alonzo Matamoros
#'
#' @seealso \code{\link{arch.test}}
#'
#' @references
#' Engle, R. F. (1982). Auto-regressive Conditional Heteroscedasticity
#' with Estimates of the Variance of United Kingdom Inflation.
#' \emph{Econometrica}. 50(4), 987-1007.
#'
#' McLeod, A. I. and W. K. Li. (1984). Diagnostic Checking ARMA Time Series
#' Models Using Squared-Residual Auto-correlations. \emph{Journal of Time
#' Series Analysis.} 4, 269-273.
#'
#' @examples
#' # generating an stationary arma process
#' y = arima.sim(100,model = list(ar = 0.3))
#' Lm.test(y)
#'
Lm.test = function(y,lag.max = 2,alpha = 0.05){

  if( !is.numeric(y) & !is(y,class2 = "ts") )
    stop("y object must be numeric or a time series")

  if( anyNA(y) )
    stop("The time series contains missing values")

  k = lag.max
  if(lag.max < 2) k = 2

  res2 =  y^2
  n = length(res2)
  if (n < 2)
    stop("not enough length of residuals")

  if(length(n) < lag.max )
    k = 2

  dname = deparse(substitute(y))
  rt = embed(res2, k)
  lm.fit = lm(rt[, 1] ~ rt[, -1])
  SSE = sum((rt[, 1] - residuals(lm.fit))^2)
  SST = sum((rt[, 1] - mean(rt[, 1]))^2)
  LM = ((SST - SSE)/k)/(SSE/(n - 2 * k - 1))
  names(LM) = "LM"
  df = k-1
  names(df) = "df"
  pval =  1 - pchisq(LM, df)

  rval = list(statistic =LM, parameter = df, p.value = pval,
              alternative = "y is heteroscedastic",
              method = "Lagrange Multiplier test", data.name = dname)

  rval$hetero = rval$p.value < alpha

  if(rval$hetero)
    rval$Conc = "Conclusion: y is heteroscedastic"
  else
    rval$Conc = "Conclusion: y is homoscedastic"

  class(rval) = "htest"
  return(rval)
}


R/rejection_rates.R
#' Rejection rates estimates
#'
#' Performs a rejection rate estimate for a stationary AR(1) process
#'
#'
#' @param reps an integer with the total bootstrap repetitions.
#' @param n an integer with the length of the simulated AR(1) process
#' @param dist optional: a function to generate the innovations. By default rnorm
#' @param seed An optional \code{\link[=set.seed]{seed}} to use.
#' @param htest A character string naming the desired test for checking normality. Valid values are
#' \code{"epps"} for the Epps, \code{"lobato"} for Lobato and Velasco's,\code{"vavra"} for the Psaradakis
#' and  Vavra,  and \code{"rp"} for the random projections. The default value is \code{"lobato"} test.
#' @param alpha  Level of the test, possible values range from 0.01 to 0.1. By default \code{alpha = 0.05}
#' is used
#' @param k an integer with the number of random projections to be used, by default
#' \code{k = 2}.
#' @param phi a real value with the autoregressive coefficent to simulate an stationary AR(1) process.
#' @param sd optional value for the standard desviation in a normal distribution.
#' @param ... additional arguments for rand.gen. Most usefully, the standard deviation of the innovations
#' generated by \code{rnorm} can be specified by \code{sd}.
#'
#' @return a real value with the proportional number of rejection for a defined test
#'
#' @noRd
#' @importFrom stats arima.sim
#'
#' @author Asael Alonzo Matamoros
#'
#' @references
#' Nieto-Reyes, A., Cuesta-Albertos, J. & Gamboa, F. (2014). A random-projection
#' based test of Gaussianity for stationary processes. \emph{Computational
#' Statistics & Data Analysis, Elsevier}, vol. 75(C), pages 124-141.
#'
#' @examples
#' # rejection rate for a Gaussian AR(1) process with phi = 0.5 and lobatos test
#' nortsTest:::rejection_rate(reps = 1,n = 100,phi = 0.5)
#'
#'
rejection_rate = function(reps = 1000,n = 100,dist = rnorm,seed = NULL,htest = "lobato",
                          alpha = 0.05,k = 64,phi = 0.5,sd = 1,...){

  if (!is.null(seed))
    set.seed(seed)

  pvalues = rep(0,reps)

  for(i in 1:reps){
    x = stats::arima.sim(n=n,n.start = 500,list(ar = phi),rand.gen = dist,...)

    if(htest == "vavra"){
      pvalues[i] = vavra.test(x,reps = 100,h = 50)$p.value
    }
    else if(htest == "lobato"){
      pvalues[i] = suppressWarnings(lobato.test(x)$p.value)
    }
    else if(htest == "epps"){
      pvalues[i] = suppressWarnings(epps.test(x)$p.value)
    }
    else if(htest == "rp"){
      pvalues[i] = suppressWarnings(rp.test(x,k = k)$p.value)
    }
  }

  rr = mean(pvalues < alpha)
  return(rr)
}
#' Rejection rates table
#'
#' Performs a rejection rate estimates tables for different stationary AR(1) processes, simulated
#' with different autorregressive coefficents and distributions, the used distributions are normal
#' \code{X ~ N(0,1)} log normal\code{X ~ logN(0,1)},t student \code{X ~ t(df = 7)}, Chi square
#' \code{X ~ Chisq(10)}, and a beta \code{X ~ beta(7,1)} distributions.
#'
#' @param reps an integer with the total bootstrap repetitions.
#' @param n an integer with the length of the simulated AR(1) process.
#' @param phi a real vectors with the autoregressive coefficents to simulate an stationary AR(1) process.
#' @param seed An optional \code{\link[=set.seed]{seed}} to use.
#' @param htest A character string naming the desired test for checking normality. Valid values are
#' \code{"epps"} for the Epps, \code{"lobato"} for Lobato and Velasco's,\code{"vavra"} for the Psaradakis
#' and  Vavra,  and \code{"rp"} for the random projections. The default value is \code{"lobato"} test.
#' @param alpha  Level of the test, possible values range from 0.01 to 0.1. By default \code{alpha = 0.05}
#' is used.
#' @param k an integer with the number of random projections to be used, by default
#' \code{k = 2}.
#'
#' @return a matrix with rejection rates estimates, where the columns represent different \code{AR(1)}
#' coefficient values, and the row different distributions.
#'
#' @noRd
#'
#' @author Asael Alonzo Matamoros
#'
#'
#' @references
#' Nieto-Reyes, A., Cuesta-Albertos, J. & Gamboa, F. (2014). A random-projection
#' based test of Gaussianity for stationary processes. \emph{Computational
#' Statistics & Data Analysis, Elsevier}, vol. 75(C), pages 124-141.
#'
#' @examples
#' # rejection rate for a Gaussian AR(1) process with phi = 0.5 and lobatos test
#' nortsTest:::rejection_table(reps = 1,n = 100)
#'
rejection_table = function(reps = 1000,n = 100,phi = c(-0.4,-0.25,0.0,0.25,0.4),
                           seed = NULL,htest = "lobato",alpha = 0.05,k = 64){

  if (!is.null(seed))
    set.seed(seed)

  nphi = length(phi)
  df = matrix(0,nrow =5,ncol = nphi)
  row.names(df) = c("N","logN","t_3","Chisq_10","beta(7,1)")
  colnames(df) = phi
  #normal
  for(i in 1:nphi) df[1,i] = rejection_rate(reps = reps,n=n,phi = phi[i],dist = rnorm,
                                         htest = htest,alpha = alpha,k = k,sd = 1,seed = seed)
  #logN
  for(i in 1:nphi) df[2,i] = rejection_rate(reps = reps,n=n,phi = phi[i],dist = rlnorm,
                                         htest = htest,alpha = alpha,k = k,sdlog = 1,seed = seed)
  #t_3
  for(i in 1:nphi) df[3,i] = rejection_rate(reps = reps,n=n,phi = phi[i],dist = rt,df = 3,
                                         htest = htest,alpha = alpha,k = k,seed = seed)
  #chi_10
  for(i in 1:nphi) df[4,i] = rejection_rate(reps = reps,n=n,phi = phi[i],dist = rchisq,df = 10,
                                         htest = htest,alpha = alpha,k = k,seed = seed)
  #beta(7,1)
  for(i in 1:nphi) df[5,i] = rejection_rate(reps = reps,n=n,phi = phi[i],dist = rbeta,shape1 = 7,shape2 = 1,
                                         htest = htest,alpha = alpha,k = k,seed = seed)
  return(df)
}


R/check_residuals.R
#'
#' @export
#'
check_residuals<- function(y,...) {
  UseMethod("check_residuals")
}
#' Generic functions for checking residuals in time series models
#'
#' Generic function for residuals check analysis, these methods are inspired in the \code{check.residuals} function
#' provided by the \code{forecast} package.
#'
#' @aliases check_residuals check_residuals.ts check_residuals.arima0 check_residuals.Arima
#' check_residuals.fGarch check_residuals.numeric check_residuals.lm  check_residuals.HoltWinters
#' check_residuals.ets check_residuals.forecast
#'
#' @rdname check_residuals
#'
#' @param y 	Either a time series model,the supported classes are \code{arima0}, \code{Arima}, \code{sarima},
#' \code{fGarch}, or a time series (assumed to be residuals).
#' @param normality A character string naming the desired test for checking gaussian distribution.
#' Valid values are \code{"epps"} for the Epps, \code{"lobato"} for Lobato and Velasco's,\code{"vavras"} for
#' the Psaradakis and  Vavra, \code{"rp"} for the random projections, \code{"jb"} for the Jarque and Beras,
#' \code{"ad"} for Anderson Darling test, and \code{"shapiro"} for the Shapiro-Wilk's test. The default value
#' is \code{"epps"} test.
#' @param unit_root A character string naming the desired unit root test for checking stationarity.
#' Valid values are \code{"adf"} for the Augmented Dickey-Fuller, \code{"pp"} for the Phillips-Perron,
#' and \code{"kpss"} for Kwiatkowski, Phillips, Schmidt, and Shin. The default value is \code{"adf"} for the
#' Augmented Dickey-Fuller test.
#' @param seasonal A character string naming the desired unit root test for checking seasonality.
#' Valid values are \code{"ocsb"} for the Osborn, Chui, Smith, and Birchenhall, \code{"ch"} for the
#' Canova and Hansen, and \code{"hegy"} for Hylleberg, Engle, Granger, and Yoo. The default value is
#' \code{"ocsb"} for the Osborn, Chui, Smith, and Birchenhall test.
#' @param arch A character string naming the desired  test for checking stationarity. Valid values are
#' \code{"box"} for the Ljung-Box, and \code{"Lm"} for the Lagrange Multiplier test. The default
#' value is \code{"box"} for the Augmented Ljung-Box test.
#' @param alpha Level of the test, possible values range from 0.01 to 0.1. By default \code{alpha = 0.05}
#' is used
#' @param plot A boolean value. If \code{TRUE}, will produce produces a time plot of the residuals,
#' the corresponding ACF, and a histogram.
#' @param ... Other testing parameters
#'
#' @details The function performs a residuals analysis, it prints a unit root and seasonal test to check
#' stationarity, and a normality test for checking  Gaussian distribution. In addition, if the plot option is
#' \code{TRUE} a time plot, ACF, and histogram of the series are presented.
#'
#' @return The function does not return any value
#'
#' @author Asael Alonzo Matamoros
#'
#' @importFrom forecast  ggtsdisplay
#' @method check_residuals ts
#'
#' @export
#'
#' @references
#' Dickey, D. & Fuller, W. (1979). Distribution of the Estimators for
#' Autoregressive Time Series with a Unit Root. \emph{Journal of the American
#' Statistical Association}. 74, 427-431.
#'
#' Epps, T.W. (1987). Testing that a stationary time series is Gaussian. \emph{The
#' Annals of Statistic}. 15(4), 1683-1698.\url{http://www.jstor.org/stable/2336512}.
#' \code{doi:10.1214/aos/1176350618}
#'
#' Osborn, D., Chui, A., Smith, J., & Birchenhall, C. (1988). Seasonality and the
#' order of integration for consumption. \emph{Oxford Bulletin of Economics
#' and Statistics}. 50(4), 361-377.
#'
#' @examples
#' # Generating an stationary arma process
#' y = arima.sim(100,model = list(ar = 0.3))
#' check_residuals(y,unit_root = "adf")
#'
check_residuals.ts = function(y,normality = "epps",
                              unit_root = NULL,
                              seasonal = NULL,
                              arch = NULL,
                              alpha = 0.05,plot = FALSE,...){

  if( !is(y,class2 = "ts") )
    stop("The object y is not a time series")

  if( anyNA(y) )
    stop("The time series contains missing values")

  # Unit root test
  if(!is.null(unit_root)){
    cc = uroot.test(y = y,unit_root = unit_root,alpha = alpha)
    cat("\n *************************************************** \n")
    cat("\n Unit root test for stationarity: \n")
    print(cc)
    cat("\n",cc$Conc)
    cat("\n *************************************************** \n")
  }

  # seasonal test
  if(frequency(y) > 1){
    if(!is.null(seasonal)){
      cc = seasonal.test(y = y,seasonal = seasonal,alpha = alpha)

      cat("\n Unit root test for seasonality: \n")
      print(cc)
      cat("\n",cc$Conc)
      cat("\n \n *************************************************** \n")
    }
  }

  # arch test
  if(!is.null(arch)){
    cc = arch.test(y = y,arch = arch,alpha = alpha,...)
    cat("\n arch test for heteroscedasticity: \n")
    print(cc)
    cat("\n",cc$Conc)
    cat("\n \n *************************************************** \n")
  }

  # Normality test
  cc = normal.test(y = y,normality =normality,alpha = alpha)
  cat("\n Goodness of fit test for Gaussian Distribution: \n")
  print(cc)
  cat("\n",cc$Conc)
  cat("\n \n *************************************************** \n")

  if (plot) {
   suppressMessages(suppressWarnings(check_plot.ts(y)))
  }
}
#'
#' @method check_residuals numeric
#' @export
#'
check_residuals.numeric = function(y,normality = "epps", unit_root = NULL,seasonal = NULL,
                                   arch = NULL,alpha = 0.05,plot = FALSE,...){
  if( !is.numeric(y) )
    stop("The object y is not a numeric array")

  check_residuals.ts(ts(y,frequency = 1),unit_root = unit_root,
                     normality = normality,
                     seasonal = seasonal,arch = arch,
                     alpha = 0.05,
                     plot = plot,...)
}
#' @method check_residuals arima0
#' @export
#'
check_residuals.arima0 = function(y,normality = "epps",
                                  unit_root = NULL,
                                  seasonal = NULL,
                                  arch = NULL,
                                  alpha = 0.05,plot = FALSE,...){
  if( !is(y,class2 = "arima0") )
    stop("The object y is not an arima0 class")

  check_residuals.ts(y$residuals,unit_root = unit_root,
                     normality = normality,
                     seasonal = seasonal,arch = arch,
                     alpha = 0.05,
                     plot = plot,...)
}
#'
#' @method check_residuals Arima
#' @export
#'
check_residuals.Arima = function(y,normality = "epps",
                                 unit_root = NULL,
                                 seasonal = NULL,
                                 arch = NULL,
                                 alpha = 0.05,plot = FALSE,...){
  if( !is(y,class2 = "Arima") )
    stop("The object y is not a Arima class")

  check_residuals.ts(y$residuals,unit_root = unit_root,
                     normality = normality,
                     seasonal = seasonal,arch = arch,
                     alpha = 0.05,
                     plot = plot,...)
}
#'
#' @method check_residuals fGARCH
#' @export
#'
check_residuals.fGARCH = function(y,normality = "epps",
                                  unit_root = NULL,
                                  seasonal = NULL,
                                  arch = NULL,
                                  alpha = 0.05,plot = FALSE,...){
  if( !is(y,class2 = "fGARCH") )
    stop("The object y is not a fGARCH class")

  check_residuals.numeric(y@residuals,unit_root = unit_root,
                          normality = normality,
                          seasonal = seasonal,arch = arch,
                          alpha = 0.05,
                          plot = plot,...)
}
#'
#' @method check_residuals lm
#' @export
#'
check_residuals.lm = function(y,normality = "epps",
                              unit_root = NULL,
                              seasonal = NULL,
                              arch = NULL,
                              alpha = 0.05,plot = FALSE,...){
  if( !is(y,class2 = "lm") )
    stop("The object y is not a lm class")

  check_residuals.numeric(y$residuals,unit_root = unit_root,
                          normality = normality,
                          seasonal = seasonal,arch = arch,
                          alpha = 0.05,
                          plot = plot,...)
}
#' @method check_residuals HoltWinters
#' @export
#'
check_residuals.HoltWinters = function(y,normality = "epps",
                                       unit_root = NULL,
                                       seasonal = NULL,
                                       arch = NULL,
                                       alpha = 0.05,plot = FALSE,...){
  if( !is(y,class2 = "HoltWinters") )
    stop("The object y is not an HoltWinters class")

  y1 = residuals(y)

  check_residuals.ts(y1,unit_root = unit_root,
                     normality = normality,
                     seasonal = seasonal,arch = arch,
                     alpha = 0.05,
                     plot = plot,...)
}
#'
#' @method check_residuals ets
#' @export
#'
check_residuals.ets = function(y,normality = "epps",
                               unit_root = NULL,
                               seasonal = NULL,
                               arch = NULL,
                               alpha = 0.05,plot = FALSE,...){
  if( !is(y,class2 = "ets") )
    stop("The object y is not a ets class")

  check_residuals.ts(y$residuals,
                     unit_root = unit_root,
                     normality = normality,
                     seasonal = seasonal,arch = arch,
                     alpha = 0.05,
                     plot = plot,...)
}
#'
#' @method check_residuals forecast
#' @export
#'
check_residuals.forecast = function(y,normality = "epps",
                                    unit_root = NULL,
                                    seasonal = NULL,
                                    arch = NULL,
                                    alpha = 0.05,plot = FALSE,...){
  if( !is(y,class2 = "forecast") )
    stop("The object y is not a forecast class")

  check_residuals.ts(y$residuals,
                     unit_root = unit_root,
                     normality = normality,
                     seasonal = seasonal,arch = arch,
                     alpha = 0.05,
                     plot = plot,...)
}
